compiler/rustc_privacy/src/lib.rs

```diff
@@ -176,6 +192,20 @@ where
 {
     type BreakTy = V::BreakTy;
 
+    fn visit_binder<T: TypeVisitable<TyCtxt<'tcx>>>(
+        &mut self,
+        t: &ty::Binder<'tcx, T>,
+    ) -> ControlFlow<Self::BreakTy> {
+        self.universes.push(None);
+        let result: ControlFlow<<V as DefIdVisitor<'tcx>>::BreakTy> = t.super_visit_with(self);
+        self.universes.pop();
+        result
+    }
+
+    fn visit_predicate(&mut self, p: ty::Predicate<'tcx>) -> ControlFlow<Self::BreakTy> {
+        self.visit_clause(p.as_clause().unwrap())
```

Comment:
The visit_clause has special logic for traits. This type visitor would not use it and thus only look at the types in the trait ref arguments, instead of discovering the trait ref's trait and marking it('s def id) as seen

Reply:
How did this work previously?
Does this change fix some pre-existing issue?

I tried to remove this change and got some missing errors.
But without this PR, on master, those errors are reported even without the `visit_predicate` override.

Reply:
FWIW, I still cannot find where the *real* default implementation of `fn visit_predicate` lives, there are too many abstraction layers.